sdks/python/apache_beam/transforms/trigger.py

```diff
@@ -1038,14 +1097,21 @@ def process_timer(self, window_id, unused_name, time_domain, timestamp,
     window = state.get_window(window_id)
     if state.get_state(window, self.TOMBSTONE):
       return
-    if time_domain == TimeDomain.WATERMARK:
-      if not self.is_merging or window in state.known_windows():
-        context = state.at(window)
+
+    if not self.is_merging or window in state.known_windows():
+      context = state.at(window, self.clock)
+      if time_domain == TimeDomain.WATERMARK:
         if self.trigger_fn.should_fire(timestamp, window, context):
           finished = self.trigger_fn.on_fire(timestamp, window, context)
           yield self._output(window, finished, state)
-    else:
-      raise Exception('Unexpected time domain: %s' % time_domain)
+      elif time_domain == TimeDomain.REAL_TIME:
+        # Fire as soon as the clock surpasses the timer's timestamp
+        if self.clock.time() > timestamp:
```

Comment:
<!--thread_id:cc_163034546_t; commit:5ddb81e7534625fa7f44e5ab7dbada5eb605bf3b; resolved:0-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Can we delegate to `self.trigger_fn.should_fire(timestamp, window, context)` in the same way?  In this case, the we would pass in the current time and have `AfterProcessingTime.should_fire()` do this time comparison.  This would make it easier for other triggers to use real time timers without assuming the trigger fires when the timer fires.

<!--section:body-->
That would assume that the timestamp is a realtime timestamp, but the call to should_fire is being made from process_elements where the timestamp is actually referring to the watermark, so it would require having the time_domain and current_time in the should_fire of AfterProcessingTime. I’m adding a todo for that change, which will include changing the interface of the parent TriggerFn.